### DISCUSSION TITLE
`Email` with null value not working properly

### DISCUSSION THREAD

**Discussion by sumitsharansatsangi** [piccolo-orm/piccolo_admin#296] [General]
My table is as follows:
```python
class Student(Table, tablename="student"):
    phone = Varchar(
        length=128,
        primary_key=False,
        index=True,
        index_method=IndexMethod.btree,
        secret=False,
    )
    first_name = Varchar(
        length=100,
        default="",
        null=True,
    )
    last_name = Varchar(
        length=100,
        default="",
        null=True,
    )
    username = Varchar(
        length=15,
        default="",
        null=False,
        primary_key=False,
    )
    email = Email(
        length=100,
        null=True,
    )
```

When I want to view the Student Table, I receive following error 
```
    json = self.pydantic_model_plural(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 3 validation errors for StudentPlural
rows -> 0 -> email
  value is not a valid email address (type=value_error.email)
rows -> 1 -> email
  value is not a valid email address (type=value_error.email)
rows -> 2 -> email
  value is not a valid email address (type=value_error.email)

```

**Comment by dantownsend** [comment]
Try setting `required=False`.

```python
email = Email(
    length=100,
    null=True,
    required=False
)
```

I don't think it's well documented - we need to improve that.

**Reply by dantownsend** [reply]
I'm away from my laptop but will try testing it later.

**Reply by dantownsend** [reply]
@sinisaos Do you think you can give it a go?

**Reply by sumitsharansatsangi** [reply]
I think, you don't need to try, Because, I was wrong. 
Earlier my table was follows:
```
class Student(Table, tablename="student"):
    phone = Varchar(
        length=128,
        primary_key=False,
        index=True,
        index_method=IndexMethod.btree,
        secret=False,
    )
    first_name = Varchar(
        length=100,
        default="",
        null=True,
    )
    last_name = Varchar(
        length=100,
        default="",
        null=True,
    )
    username = Varchar(
        length=15,
        default="",
        null=False,
        primary_key=False,
    )
    email = Email(
        length=100,
        null=True,
       default="",   # This line creates the problem
    )
```
I changed to following
```
class Student(Table, tablename="student"):
    phone = Varchar(
        length=128,
        primary_key=False,
        index=True,
        index_method=IndexMethod.btree,
        secret=False,
    )
    first_name = Varchar(
        length=100,
        default="",
        null=True,
    )
    last_name = Varchar(
        length=100,
        default="",
        null=True,
    )
    username = Varchar(
        length=15,
        default="",
        null=False,
        primary_key=False,
    )
    email = Email(
        length=100,
        null=True,
    )
```
But the error persists.
Then I did message, here. 
Later, I find , I have just confused empty with null values.
All has to be null or valid email, but all has empty values.  

![Screenshot from 2023-04-04 15-58-32](https://user-images.githubusercontent.com/45959281/229764936-10eef30d-f066-4bad-995b-8e097b4c0ef1.png)

EmailField is working fine, no issue there.
I was wrong.

**Comment by sumitsharansatsangi** [comment]
@dantownsend , I want to draw your attention in this issue again. 
What `Email` is doing is that it inherits all the property from `Varchar`.  And what `Varchar` is doing is that  when it sees null  value , it substitute it with Empty String , i.e. "" which is not a valid email. And for `Email` when admin is checking for valid email. it gives error.

I don't know exactly how it can be fixed. 
One way to do is admin should check for valid email before assigning the empty string i.e. "". But I think this requires implementation of `Email` from scratch, not by just inheriting `Varchar`. 

@sinisaos , can you suggest something....?

**Comment by sinisaos** [comment]
@sumitsharansatsangi Quick fix will be change [this line in Piccolo](https://github.com/piccolo-orm/piccolo/blob/87dc9de542a3c5c39b194610a1df94d27f292da2/piccolo/utils/pydantic.py#L221-L222) to this:
```python
elif isinstance(column, Email):
    value_type = t.Union[pydantic.EmailStr, t.Literal[""]]
```
This would allow an empty string to pass validation. But that needs to be checked, but I don't have time right now so I'll have a look tomorrow.

**Reply by sumitsharansatsangi** [reply]
I'll check it, @sinisaos . Thanks for the quick reply.
No, problem , we will see it tomorrow, if any problem occurs.

**Reply by sumitsharansatsangi** [reply]
Thanks @sinisaos . It passes the validation and working fine.

**Reply by sinisaos** [reply]
@sumitsharansatsangi Actually, I think a much better solution (nothing changes in Piccolo ORM) would be to add this 
```javascript
else if (schema?.properties[key].format == "email" && value == "") {
    value = null
}
```
to [utils.ts](https://github.com/piccolo-orm/piccolo_admin/blob/ed65c66deafef2ba6e6916587d65e836b8b8fd33/admin_ui/src/utils.ts#L98) (just like the other fields) and convert form value in Piccolo Admin so that the `null` Email fields work correctly. I just tried that and it works. @dantownsend what do you think about that?

**Reply by dantownsend** [reply]
> to [utils.ts](https://github.com/piccolo-orm/piccolo_admin/blob/ed65c66deafef2ba6e6916587d65e836b8b8fd33/admin_ui/src/utils.ts#L98) (just like the other fields) and convert form value in Piccolo Admin so that the null Email fields work correctly. I just tried that and it works. @dantownsend what do you think about that?

@sinisaos Yes, this is the best fix.

I tried out a quick demo project:

```python
from fastapi import FastAPI
from piccolo_admin.endpoints import create_admin
from piccolo.apps.user.tables import BaseUser
from piccolo_api.session_auth.tables import SessionsBase
from piccolo.table import Table, create_db_tables
from piccolo.columns import Email


class Manager(Table):
    email = Email(null=True, default=None, required=False)


app = FastAPI()


app.mount('/admin/', create_admin(tables=[Manager]))


@app.on_event('startup')
async def startup():
    await create_db_tables(BaseUser, SessionsBase, Manager, if_not_exists=True)
```

And yeah, it gets a validation error as it's posting back an empty string.

<img width="685" alt="Screenshot 2023-05-30 at 08 30 21" src="https://github.com/piccolo-orm/piccolo_admin/assets/350976/e2ef7c05-a5d1-4d1a-b9ce-f81c5dddcaa9">

As you identified, we use `convertFormValue` in `utils.ts` to clean up values before sending them to the API.

Let's get this fix in, so `convertFormValue` concerts an empty string to `null` for email columns.

Longer term we need to think more about how null values are handled in Piccolo Admin. We might need an explicit null checkbox or something.

<img width="676" alt="Screenshot 2023-05-30 at 08 43 07" src="https://github.com/piccolo-orm/piccolo_admin/assets/350976/4d2ad747-73a8-456b-9341-786010963d71">

We could just automatically convert all empty strings to null, but Postgres does allow both, and they're not the same thing. Here's some [discussion about it](https://dba.stackexchange.com/questions/59/when-to-use-null-and-when-to-use-an-empty-string).

But yeah, lets get this fix in for now.

**Reply by sinisaos** [reply]
@dantownsend Great. I can do a quick PR with that fix and some tests if you want.

**Reply by dantownsend** [reply]
@sinisaos Thanks, if you could then that would be great, and I'll get it merged in asap.

**Reply by dantownsend** [reply]
This should be fixed now in `piccolo_admin==0.50.0`.

**Reply by sumitsharansatsangi** [reply]
No @dantownsend , `piccolo_admin==0.50.0` doesn't fixed the error.  What fixed the error is the following answer given by  @sinisaos .  Does this answer has any side effect, or may raise any problem in future?

> @sumitsharansatsangi Quick fix will be change [this line in Piccolo](https://github.com/piccolo-orm/piccolo/blob/87dc9de542a3c5c39b194610a1df94d27f292da2/piccolo/utils/pydantic.py#L221-L222) to this:
> 
> ```python
> elif isinstance(column, Email):
>     value_type = t.Union[pydantic.EmailStr, t.Literal[""]]
> ```
> 
> This would allow an empty string to pass validation. But that needs to be checked, but I don't have time right now so I'll have a look tomorrow.

**Reply by dantownsend** [reply]
Does your column look like this:

```python
Email(null=True, default=None, required=False)
```

Changing the Pydantic validation works, but isn't the root cause, so would rather not change it.

**Reply by sinisaos** [reply]
@sumitsharansatsangi I don't know what is causing your problems, but all options work for me
```python
email = Email(null=True)
email = Email(null=True, default="")
email = Email(null=True, default=None)
email = Email(null=True, default=None, required=False)
```
@dantownsend All options works for me

**Reply by dantownsend** [reply]
Maybe the files are cached in the browser? I'm not at my laptop currently, so can't check.

**Reply by sumitsharansatsangi** [reply]
@sinisaos, As @dantownsend said, I think cache was the problem.
It's working now, you don't need to check @dantownsend . 
Thanks...
